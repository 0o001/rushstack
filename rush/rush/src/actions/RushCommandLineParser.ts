/**
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 */

import * as os from 'os';
import * as colors from 'colors';

import { CommandLineFlagParameter } from '../commandLine/CommandLineParameter';
import CommandLineParser from '../commandLine/CommandLineParser';
import RushConfig from '../data/RushConfig';
import LinkAction from './LinkAction';
import RebuildAction from './RebuildAction';
import PrepareAction from './PrepareAction';

export default class RushCommandLineParser extends CommandLineParser {
  public rushConfig: RushConfig;
  private _debugParameter: CommandLineFlagParameter;

  constructor() {
    super({
      toolFilename: 'rush',
      toolDescription: 'Rush helps you to manage a collection of Node Package Manager'
      + ' projects.  (For more details about NPM, see the www.npmjs.com web site.)'
      + ' Rush collects the dependencies for all projects into a "common" folder,'
      + ' detects which projects can be locally linked, and then performs a fast parallel'
      + ' build according to the detected dependency graph.  If you want to decompose'
      + ' your monolithic project into many small packages but are afraid of the dreaded'
      + ' NPM progress bar, Rush is for you.'
    });

    this.addAction(new PrepareAction(this));
    this.addAction(new LinkAction(this));
    this.addAction(new RebuildAction(this));
  }

  protected onDefineParameters(): void {
    this._debugParameter = this.defineFlagParameter({
      parameterLongName: '--debug',
      parameterShortName: '-d',
      description: 'Show the full call stack if an error occurs while executing the tool'
    });
  }

  protected onExecute(): void {
    this.trapErrors(() => {
      super.onExecute();
    });
  }

  public trapErrors(action: () => void): void {
    if (this._debugParameter.value) {
      action();
    } else {
      try {
        action();
      } catch (error) {
        console.error(os.EOL + colors.red('ERROR: ' + error.message));
        process.exit(1);
      }
    }
  }

}
