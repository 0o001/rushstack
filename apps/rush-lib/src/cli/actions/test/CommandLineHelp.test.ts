// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as colors from 'colors';
import { RushCommandLineParser } from '../RushCommandLineParser';

describe('CommandLineHelp', () => {
  const parser: RushCommandLineParser = new RushCommandLineParser();

  it('prints the global help', () => {
    const helpText: string = colors.stripColors(parser.renderHelpText());
    expect(helpText).toMatchSnapshot();
  });

  for (const action of parser.actions) {
    it(`prints the help for "rush ${action.actionName}"`, () => {
      const helpText: string = colors.stripColors(action.renderHelpText());
      expect(helpText).toMatchSnapshot();
    });
  }

});
