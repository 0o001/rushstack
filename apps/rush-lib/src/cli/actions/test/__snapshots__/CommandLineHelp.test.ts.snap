// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandLineHelp prints the global help 1`] = `
"usage: rush [-h] [-d] <command> ...

Rush helps you to manage a collection of npm projects. Rush collects the 
dependencies for all projects to perform a minimal install, detects which 
projects can be locally linked, and performs a fast parallel build according 
to the detected dependency graph. If you want to decompose your monolithic 
project into many small packages but are afraid of the dreaded NPM progress 
bar, Rush is for you.

Positional arguments:
  <command>
    change     Records changes made to projects, indicating how the package 
               version number should be bumped for the next publish.
    check      Checks each project's package.json files and ensures that all 
               dependencies are of the same version throughout the repository.
    install
    link       Create node_modules symlinks for all projects
    publish    Reads and processes package publishing change requests 
               generated by \\"rush change\\".
    purge      For diagnostic purposes, use this command to delete caches and 
               other temporary files used by Rush
    scan       Scan the current project folder and display a report of 
               imported packages.
    update
    unlink     Delete node_modules symlinks for all projects
    version    (EXPERIMENTAL) Manage package versions in the repo.
    build      (EXPERIMENTAL) Build all projects that haven't been built, or 
               have changed since they were last built.
    rebuild    Clean and rebuild the entire set of projects

Optional arguments:
  -h, --help   Show this help message and exit.
  -d, --debug  Show the full call stack if an error occurs while executing 
               the tool

For detailed help about a specific command, use: rush <command> -h
"
`;

exports[`CommandLineHelp prints the help for "rush build" 1`] = `
"usage: rush build [-h] [-p COUNT] [-t PROJECT1]
                  [--to-version-policy VERSION_POLICY_NAME] [-f PROJECT2] [-v]
                  [-o] [--no-color] [--production]
                  

The Rush build command assumes that the package.json file for each project 
contains scripts for \\"npm run clean\\" and \\"npm run test\\". It invokes these 
commands to build each project. Projects are built in parallel where possible,
 but always respecting the dependency graph for locally linked projects. The 
number of simultaneous processes will be based on the number of machine cores 
unless overridden by the --parallelism flag.

Optional arguments:
  -h, --help            Show this help message and exit.
  -p COUNT, --parallelism COUNT
                        Specify the number of concurrent build processes The 
                        value \\"max\\" can be specified to indicate the number 
                        of CPU cores. If this parameter omitted, the default 
                        value depends on the operating system and number of 
                        CPU cores.
  -t PROJECT1, --to PROJECT1
                        Run command in the specified project and all of its 
                        dependencies
  --to-version-policy VERSION_POLICY_NAME
                        Run command in all projects with the specified 
                        version policy and all of their dependencies
  -f PROJECT2, --from PROJECT2
                        Run command in all projects that directly or 
                        indirectly depend on the specified project
  -v, --verbose         Display the logs during the build, rather than just 
                        displaying the build status summary
  -o, --changed-projects-only
                        If specified, the incremental build will only rebuild 
                        projects that have changed, but not any projects that 
                        directly or indirectly depend on the changed package.
  --no-color            disable colors in the build log, defaults to 'true'
  --production          Perform a production build, including minification 
                        and localization steps
"
`;

exports[`CommandLineHelp prints the help for "rush change" 1`] = `
"usage: rush change [-h] [-v] [-b BRANCH]

Asks a series of questions and then generates a <branchname>-<timstamp>.json 
file in the common folder. The \`publish\` command will consume these files and 
perform the proper version bumps. Note these changes will eventually be 
published in a changelog.md file in each package. The possible types of 
changes are: MAJOR - these are breaking changes that are not backwards 
compatible. Examples are: renaming a public class, adding/removing a 
non-optional parameter from a public API, or renaming an variable or function 
that is exported. MINOR - these are changes that are backwards compatible 
(but not forwards compatible). Examples are: adding a new public API or 
adding an optional parameter to a public API PATCH - these are changes that 
are backwards and forwards compatible. Examples are: Modifying a private API 
or fixing a bug in the logic of how an existing API works. HOTFIX 
(EXPERIMENTAL) - these are changes that are hotfixes targeting a specific 
older version of the package. When a hotfix change is added, other changes 
will not be able to increment the version number.Enable this feature by 
setting 'hotfixChangeEnabled' in your rush.json.

Optional arguments:
  -h, --help            Show this help message and exit.
  -v, --verify          Verify the change file has been generated and that it 
                        is a valid JSON file
  -b BRANCH, --target-branch BRANCH
                        If this parameter is specified, compare current 
                        branch with the target branch to get changes. If this 
                        parameter is not specified, the current branch is 
                        compared against the \\"master\\" branch.
"
`;

exports[`CommandLineHelp prints the help for "rush check" 1`] = `
"usage: rush check [-h]

Checks each project's package.json files and ensures that all dependencies 
are of the same version throughout the repository.

Optional arguments:
  -h, --help  Show this help message and exit.
"
`;

exports[`CommandLineHelp prints the help for "rush install" 1`] = `
"usage: rush install [-h] [-p] [--bypass-policy] [--no-link]

Optional arguments:
  -h, --help       Show this help message and exit.
  -p, --purge
  --bypass-policy  Overrides \\"gitPolicy\\" enforcement (use honorably!)
  --no-link
"
`;

exports[`CommandLineHelp prints the help for "rush link" 1`] = `
"usage: rush link [-h] [-f]

Create node_modules symlinks for all projects

Optional arguments:
  -h, --help   Show this help message and exit.
  -f, --force  Deletes and recreates all links, even if the filesystem state 
               seems to indicate that this is unnecessary.
"
`;

exports[`CommandLineHelp prints the help for "rush publish" 1`] = `
"usage: rush publish [-h] [-a] [-b BRANCH] [-p] [--add-commit-details]
                    [--regenerate-changelogs] [-r REGISTRY] [-n TOKEN]
                    [-t TAG] [--pack] [--release-folder FOLDER]
                    [--release-type RELEASE_TYPE] [--include-all]
                    [--version-policy POLICY] [--prerelease-name NAME]
                    [--suffix SUFFIX] [--force]
                    

Reads and processes package publishing change requests generated by \\"rush 
change\\". This will perform a read-only operation by default, printing 
operations executed to the console. To commit changes and publish packages, 
you must use the --commit flag and/or the --publish flag.

Optional arguments:
  -h, --help            Show this help message and exit.
  -a, --apply           If this flag is specified, the change requests will 
                        be applied to package.json files.
  -b BRANCH, --target-branch BRANCH
                        If this flag is specified, applied changes and 
                        deleted change requests will becommitted and merged 
                        into the target branch.
  -p, --publish         If this flag is specified, applied changes will be 
                        published to npm.
  --add-commit-details  Adds commit author and hash to the changelog.json 
                        files for each change.
  --regenerate-changelogs
                        Regenerates all changelog files based on the current 
                        JSON content.
  -r REGISTRY, --registry REGISTRY
                        Publishes to a specified NPM registry. If this is 
                        specified, it will prevent the current commit will 
                        not be tagged.
  -n TOKEN, --npm-auth-token TOKEN
                        Provide the default scope NPM auth token to be passed 
                        into npm publish for global package publishing.
  -t TAG, --tag TAG     The tag option to pass to npm publish. By default NPM 
                        will publish using the 'latest' tag, even if the 
                        package is older than the current latest, so in 
                        publishing workflows for older releases, providing a 
                        tag is important. When hotfix changes are made, this 
                        parameter defaults to 'hotfix'.
  --pack                Packs projects into tarballs instead of publishing to 
                        npm repository. It can only be used when 
                        --include-all is specified. If this flag is specified,
                         NPM registry related parameters will be ignored.
  --release-folder FOLDER
                        This parameter is used with --pack parameter to 
                        provide customized location for the tarballs instead 
                        of the default value.
  --release-type RELEASE_TYPE
                        This parameter is used with --pack parameter to 
                        provide release type for the generated tarballs. The 
                        default value is 'internal'. The valid values include 
                        'public', 'beta', 'internal'
  --include-all         If this flag is specified, all packages with 
                        shouldPublish=true in rush.json or with a specified 
                        version policy will be published if their version is 
                        newer than published version.
  --version-policy POLICY
                        Version policy name. Only projects with this version 
                        policy will be published if used with --include-all.
  --prerelease-name NAME
                        Bump up to a prerelease version with the provided 
                        prerelease name. Cannot be used with --suffix
  --suffix SUFFIX       Append a suffix to all changed versions. Cannot be 
                        used with --prerelease-name.
  --force               If this flag is specified with --publish, packages 
                        will be published with --force on npm
"
`;

exports[`CommandLineHelp prints the help for "rush purge" 1`] = `
"usage: rush purge [-h] [--unsafe]

The \\"rush purge\\" command is used to delete temporary files created by Rush.

Optional arguments:
  -h, --help  Show this help message and exit.
  --unsafe    (UNSAFE!) Also delete shared files such as the package manager 
              instances stored in the \\".rush\\" folder in the user's home 
              directory. This is a more aggressive fix that is NOT SAFE to 
              run in a live environment because it will cause other 
              concurrent Rush processes to fail.
"
`;

exports[`CommandLineHelp prints the help for "rush rebuild" 1`] = `
"usage: rush rebuild [-h] [-p COUNT] [-t PROJECT1]
                    [--to-version-policy VERSION_POLICY_NAME] [-f PROJECT2]
                    [-v] [--no-color] [--production]
                    

The Rush build command assumes that the package.json file for each project 
contains scripts for \\"npm run clean\\" and \\"npm run test\\". It invokes these 
commands to build each project. Projects are built in parallel where possible,
 but always respecting the dependency graph for locally linked projects. The 
number of simultaneous processes will be based on the number of machine cores 
unless overridden by the --parallelism flag.

Optional arguments:
  -h, --help            Show this help message and exit.
  -p COUNT, --parallelism COUNT
                        Specify the number of concurrent build processes The 
                        value \\"max\\" can be specified to indicate the number 
                        of CPU cores. If this parameter omitted, the default 
                        value depends on the operating system and number of 
                        CPU cores.
  -t PROJECT1, --to PROJECT1
                        Run command in the specified project and all of its 
                        dependencies
  --to-version-policy VERSION_POLICY_NAME
                        Run command in all projects with the specified 
                        version policy and all of their dependencies
  -f PROJECT2, --from PROJECT2
                        Run command in all projects that directly or 
                        indirectly depend on the specified project
  -v, --verbose         Display the logs during the build, rather than just 
                        displaying the build status summary
  --no-color            disable colors in the build log, defaults to 'true'
  --production          Perform a production build, including minification 
                        and localization steps
"
`;

exports[`CommandLineHelp prints the help for "rush scan" 1`] = `
"usage: rush scan [-h]

The NPM system allows a project to import dependencies without explicitly 
listing them in its package.json file. This is a dangerous practice, because 
there is no guarantee you will get a compatible version. The \\"rush scan\\" 
command reports a list of packages that are imported by your code, which you 
can compare against your package.json file to find mistakes. It searches the 
\\"./src\\" and \\"./lib\\" folders for typical import syntaxes such as \\"import __ 
from '__'\\", \\"require('__')\\", \\"System.import('__'), etc. The results are only 
approximate, but generally pretty accurate.

Optional arguments:
  -h, --help  Show this help message and exit.
"
`;

exports[`CommandLineHelp prints the help for "rush unlink" 1`] = `
"usage: rush unlink [-h]

This removes the symlinks created by the \\"rush link\\" command. This is useful 
for cleaning a repo using \\"git clean\\" without accidentally deleting source 
files, or for using standard npm commands on a project.

Optional arguments:
  -h, --help  Show this help message and exit.
"
`;

exports[`CommandLineHelp prints the help for "rush update" 1`] = `
"usage: rush update [-h] [-p] [--bypass-policy] [--no-link] [--full]
                   [--recheck]
                   

Optional arguments:
  -h, --help       Show this help message and exit.
  -p, --purge
  --bypass-policy  Overrides \\"gitPolicy\\" enforcement (use honorably!)
  --no-link
  --full
  --recheck
"
`;

exports[`CommandLineHelp prints the help for "rush version" 1`] = `
"usage: rush version [-h] [-b BRANCH] [--ensure-version-policy]
                    [--override-version NEW_VERSION] [--bump]
                    [--bypass-policy] [--version-policy POLICY]
                    [--override-bump BUMPTYPE] [--override-prerelease-id ID]
                    

(EXPERIMENTAL) use this \\"rush version\\" command to ensure version policies and 
bump versions.

Optional arguments:
  -h, --help            Show this help message and exit.
  -b BRANCH, --target-branch BRANCH
                        If this flag is specified, changes will be committed 
                        and merged into the target branch.
  --ensure-version-policy
                        Updates package versions if needed to satisfy version 
                        policies.
  --override-version NEW_VERSION
                        Override the version in the specified 
                        --version-policy. This setting only works for 
                        lock-step version policy and when 
                        --ensure-version-policy is specified.
  --bump                Bumps package version based on version policies.
  --bypass-policy       Overrides \\"gitPolicy\\" enforcement (use honorably!)
  --version-policy POLICY
                        The name of the version policy
  --override-bump BUMPTYPE
                        Overrides the bump type in the version-policy.json 
                        for the specified version policy.Valid BUMPTYPE 
                        values include: prerelease, patch, preminor, minor, 
                        major. This setting only works for lock-step version 
                        policy in bump action.
  --override-prerelease-id ID
                        Overrides the prerelease identifier in the version 
                        value of version-policy.json for the specified 
                        version policy. This setting only works for lock-step 
                        version policy. This setting increases to new 
                        prerelease id when \\"--bump\\" is provided but only 
                        replaces the prerelease name when 
                        \\"--ensure-version-policy\\" is provided.
"
`;
